Automattic/wp-calypso
# Plans 2023: Extract and refactor isPlanUpgradeCreditsVisible

Fixes https://github.com/Automattic/wp-calypso/issues/78998

## Proposed Changes

This PR does a couple of things (which will lead to further refactors once merged) 

- Migrates `useIsPlanUpgradeCreditsVisible` hook out of the plans-grid component to uncouple grid from state dependencies
- Refactors `useIsPlanUpgradeCreditsVisible` to `useProRatedCreditsApplicable` as a better source for singly handling upgrade credits (`null` if not applicable, `number` if present and applicable).

### TODOs, considerations

- "being visible or not" (like previously) does give a more specific semantic, as these credits are not applicable to any plan. **this is something to consider when approving or not the changes**.
- we could feed the value through `pricing meta` instead of a prop on the grid
    - `pricing meta` is per plan, so not sure if it will make sense or not lead to weird complexities

## Testing Instructions



May need to revise this into a more thorough test plan as there may be a few more things to test. These should be applicable:

* go to `/plans[paid site]`
* ensure pro-rated credits are displayed in the notice
* ensure badges read "credits applied"
* go to `/plans/[free]` with a currency that brings up currency discounts (INR)
* ensure badges read "one time discount"
* unit tests, for the win

## Pre-merge Checklist
